File: example-codes-py/stock-brokerage-system/member.py
from datetime import datetime
from abc import ABC
from .constants import OrderStatus, AccountStatus, ReturnStatus
from .order import LimitOrder
from .stock_exchange import StockExchange


class Account(ABC):
    def __init__(self, id, password, name, address, email, phone, status=AccountStatus.NONE):
        self.__id = id
        self.__password = password
        self.__name = name
        self.__address = address
        self.__email = email
        self.__phone = phone
        self.__status = AccountStatus.NONE

    def reset_password(self):
        None


class Member(Account):
    def __init__(self):
        self.__available_funds_for_trading = 0.0
        self.__date_of_membership = datetime.date.today()
        self.__stock_positions = {}
        self.__active_orders = {}

    def place_sell_limit_order(self, stock_id, quantity, limit_price, enforcement_type):
        # check if member has this stock position
        if stock_id not in self.__stock_positions:
            return ReturnStatus.NO_STOCK_POSITION

        stock_position = self.__stock_positions[stock_id]
        # check if the member has enough quantity available to sell
        if stock_position.get_quantity() < quantity:
            return ReturnStatus.INSUFFICIENT_QUANTITY

        order = LimitOrder(stock_id, quantity, limit_price, enforcement_type)
        order.is_buy_order = False
        order.save_in_DB()
        success = StockExchange.place_order(order)
        if success:
            order.set_status(OrderStatus.FAILED)
            order.save_in_DB()
        else:
            self.active_orders.add(order.get_order_id(), order)
        return success

    def place_buy_limit_order(self, stock_id, quantity, limit_price, enforcement_type):
        # check if the member has enough funds to buy this stock
        if self.__available_funds_for_trading < quantity * limit_price:
            return ReturnStatus.INSUFFICIENT_FUNDS

        order = LimitOrder(stock_id, quantity, limit_price, enforcement_type)
        order.is_buy_order = True
        order.save_in_DB()
        success = StockExchange.place_order(order)
        if not success:
            order.set_status(OrderStatus.FAILED)
            order.save_in_DB()
        else:
            self.active_orders.add(order.get_order_id(), order)
        return success

    # this function will be invoked whenever there is an update from
    # stock exchange against an order
    def callback_stock_exchange(self, order_id, order_parts, status):
        order = self.active_orders[order_id]
        order.add_order_parts(order_parts)
        order.set_status(status)
        order.update_in_DB()

        if status == OrderStatus.FILLED or status == OrderStatus.CANCELLEd:
            self.active_orders.remove(order_id)

File: example-codes/stock-brokerage-system/java/Member.java
package com.stockbrokerage;

import java.util.HashMap;
import java.util.Map;

/**
 * Represents a brokerage account member (trader/investor).
 * 
 * Each member can place buy/sell orders through the StockExchange.
 * The class maintains available funds, owned stock positions,
 * and currently active orders.
 */
public class Member extends Account {
    private double availableFunds;
    private Map<String, Integer> stockPositions;  // Stock ID → quantity owned
    private Map<String, Order> activeOrders;      // Order ID → order object

    public Member(String id, String password, String name, Constants.Location address, String email, String phone) {
        super(id, password, name, address, email, phone, Constants.AccountStatus.ACTIVE);
        this.availableFunds = 50000.0;  // Starting balance for simplicity
        this.stockPositions = new HashMap<>();
        this.activeOrders = new HashMap<>();
    }

    /**
     * Places a BUY order with a price limit.
     * Checks whether the member has enough balance to execute.
     */
    public Constants.ReturnStatus placeBuyLimitOrder(String stockId, int quantity, double priceLimit) {
        double totalCost = quantity * priceLimit;

        if (availableFunds < totalCost) {
            System.out.println("Not enough funds to buy " + stockId);
            return Constants.ReturnStatus.INSUFFICIENT_FUNDS;
        }

        Order order = new LimitOrder(stockId + "_BUY", quantity, priceLimit, true);
        order.saveInDB();
        Constants.ReturnStatus result = StockExchange.getInstance().placeOrder(order);

        if (result == Constants.ReturnStatus.SUCCESS) {
            availableFunds -= totalCost; // Deduct funds after placing
            activeOrders.put(order.orderId, order);
        }
        return result;
    }

    /**
     * Places a SELL order for a stock the user owns.
     * Validates if they have enough quantity to sell.
     */
    public Constants.ReturnStatus placeSellLimitOrder(String stockId, int quantity, double priceLimit) {
        if (!stockPositions.containsKey(stockId)) {
            System.out.println("Cannot sell " + stockId + " — no holdings found.");
            return Constants.ReturnStatus.NO_STOCK_POSITION;
        }

        int ownedQty = stockPositions.get(stockId);
        if (ownedQty < quantity) {
            System.out.println("Not enough quantity to sell " + stockId);
            return Constants.ReturnStatus.INSUFFICIENT_QUANTITY;
        }

        Order order = new LimitOrder(stockId + "_SELL", quantity, priceLimit, false);
        order.saveInDB();
        Constants.ReturnStatus result = StockExchange.getInstance().placeOrder(order);

        if (result == Constants.ReturnStatus.SUCCESS) {
            stockPositions.put(stockId, ownedQty - quantity);
            activeOrders.put(order.orderId, order);
        }
        return result;
    }
}


/**
 * Abstract base class for all types of user accounts.
 * 
 * Handles common attributes such as credentials, name, address, and contact info.
 */
abstract class Account {
    protected String id, password, name, email, phone;
    protected Constants.Location address;
    protected Constants.AccountStatus status;

    public Account(String id, String password, String name, Constants.Location address,
                   String email, String phone, Constants.AccountStatus status) {
        this.id = id;
        this.password = password;
        this.name = name;
        this.address = address;
        this.email = email;
        this.phone = phone;
        this.status = status;
    }

    /**
     * Simulates a password reset.
     */
    public void resetPassword() {
        System.out.println("Password reset link sent to: " + email);
    }
}
